presto-spi/src/main/java/com/facebook/presto/spi/type/Type.java

```diff
@@ -141,4 +142,9 @@
      * Compare the values in the specified block at the specified positions equal.
      */
     int compareTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition);
+
+    /**
+     * Get the encoding for this type.
+     */
+    BlockEncoding getEncoding();
```

Comment:
The right thing to do is to allow any getJavaType() to return any value (or at least any Object type to start with).

In the short term, the details of how a structural type is encoded is private to the structural types and the helper methods (or it should be private).  That means you can just choose any block type and encoding you want.  For example, if you want to encode map as `[key-block][value-block]` where the block types are `FixedWidthBlock` for instances of `FixedType` and `VariableWitdthBlock` for everything else, you can simple choose to do that.  You don't need to use the `BlockBuilder` from `Type.createBlockBuilder`.

That said, I would suggest we come up with a good enough format and "lock" it down, so we don't have to keep changing all of the connectors.  In the long term, we need to allow any type on the stack (`getJavaType()` should allow any Java class).  Additionally the structural types should be encoded as columns like ORC does internally (that is a longer discussion).